libseam_carving/carver.cpp

```diff
@@ -6,239 +6,247 @@
-                } else { // seam inserts (current_coord.row, col)
-                    // get top partition
-                    const cv::Mat &top_partition = current_src_col.rowRange(0, current_coord.row());
-                    // get bottom partition
-                    const cv::Mat &bottom_partition = current_src_col.rowRange(current_coord.row(), in_mat.rows);
-                    // insert top, new, middle
-                    mats.push_back(top_partition);
-                    mats.push_back(row_insert);
-                    mats.push_back(bottom_partition);
-
-                    // color correct since i will have top and bottom neighbors
-                    const cv::Vec3b &top = in_mat.at<cv::Vec3b>(current_coord.row(), current_coord.col());
-                    const cv::Vec3b &bottom = in_mat.at<cv::Vec3b>(current_coord.row() + 1, current_coord.col());
-                    row_insert.at<cv::Vec3b>(0, 0) = (top + bottom) / 2;
-                }
+        res.copyTo(output);
+    }
+
+    void Carver::InsertHorizontalSeam(const Seam &seam, cv::InputArray input, cv::OutputArray output) {
+        cv::Mat in_mat = input.getMat();
+        cv::Mat res = cv::Mat(in_mat.rows + 1, in_mat.cols, in_mat.type());
+
+        // loop through all cols
+        for (int idx = 0; idx < res.cols; idx++) {
+            // grab the current row and coord
+            const cv::Mat &current_src_col = in_mat.col(idx);
+            const Coord &current_coord = seam[idx];
+            std::vector<cv::Mat> mats;
+
+            // create a 1x1 to insert with the pixel in the input
+            cv::Mat row_insert = cv::Mat(
+                    1, 1, in_mat.type(),
+                    in_mat.at<cv::Vec3b>(current_coord.row(), current_coord.col())
+            );
+
+            // check order to merge since you can't have a jagged matrix
+            if (current_coord.row() == 0) { // seam inserts (0, col)
+                // insert new row then original col
+                mats.push_back(row_insert);
+                mats.push_back(current_src_col);
+            } else if (current_coord.row() == in_mat.rows - 1) { // seam inserts (in_mat.rows - 1, col)
+                // insert original col then new row
+                mats.push_back(current_src_col);
+                mats.push_back(row_insert);
+            } else { // seam inserts (current_coord.row, col)
+                // get top partition
+                const cv::Mat &top_partition = current_src_col.rowRange(0, current_coord.row()).clone();
+                // get bottom partition
+                const cv::Mat &bottom_partition = current_src_col.rowRange(current_coord.row(), in_mat.rows).clone();
+                // insert top, new, middle
+                mats.push_back(top_partition);
+                mats.push_back(row_insert);
+                mats.push_back(bottom_partition);
+
+                // color correct since i will have top and bottom neighbors
```

Comment:
color correction commented out, for now, to make these easier on testing, will be added in post v2